go/libraries/doltcore/sqle/cluster/branch_control_replica.go

```diff
@@ -0,0 +1,251 @@
+			}()
+			continue
+		}
+		r.backoff.Reset()
+		r.lgr.Debugf("branchControlReplica[%s]: sucessfully replicated branch control permissions.", r.client.remote)
+		r.replicatedVersion = r.version
+	}
+}
+
+func (r *branchControlReplica) wait() {
+	r.cond.Wait()
+}
+
+func (r *branchControlReplica) isCaughtUp() bool {
+	return r.version == r.replicatedVersion || r.role != RolePrimary
+}
+
+func (r *branchControlReplica) setWaitNotify(notify func()) bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if notify != nil {
+		if r.waitNotify != nil {
+			return false
+		}
+		notify()
+	}
+	r.waitNotify = notify
+	return true
+}
+
+func (r *branchControlReplica) GracefulStop() {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.shutdown = true
+	r.cond.Broadcast()
+}
+
+func (r *branchControlReplica) setRole(role Role) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.role = role
+	r.nextAttempt = time.Time{}
+	r.cond.Broadcast()
+}
+
+func (p *branchControlReplication) setRole(role Role) {
+	if role == RolePrimary {
+		cur := p.bcController.Serialized.Load()
+		if cur == nil {
+			p.UpdateBranchControlContents([]byte{})
+		} else {
+			p.UpdateBranchControlContents(*cur)
+		}
+	}
+	for _, r := range p.replicas {
+		r.setRole(role)
+	}
+}
+
+func (p *branchControlReplication) Run() {
+	var wg sync.WaitGroup
+	for _, r := range p.replicas {
+		r := r
+		wg.Add(1)
+		func() {
+			defer wg.Done()
+			r.Run()
+		}()
+	}
+	wg.Wait()
+}
+
+func (p *branchControlReplication) GracefulStop() {
+	for _, r := range p.replicas {
+		r.GracefulStop()
+	}
+}
+
+func (p *branchControlReplication) UpdateBranchControlContents(contents []byte) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.current = contents
+	p.version += 1
+	for _, r := range p.replicas {
+		r.UpdateContents(p.current, p.version)
+	}
+}
+
+func (p *branchControlReplication) waitForReplication(timeout time.Duration) bool {
+	p.mu.Lock()
+	replicas := make([]*branchControlReplica, len(p.replicas))
+	copy(replicas, p.replicas)
+	caughtup := make([]bool, len(replicas))
+	var wg sync.WaitGroup
+	wg.Add(len(replicas))
+	for li, lr := range replicas {
+		i := li
+		r := lr
+		ok := r.setWaitNotify(func() {
+			// called with r.mu locked.
+			if !caughtup[i] {
+				if r.isCaughtUp() {
+					caughtup[i] = true
+					wg.Done()
+				} else {
+				}
+			}
+		})
+		if !ok {
+			for j := li - 1; j >= 0; j-- {
+				replicas[j].setWaitNotify(nil)
+			}
+			return false
+		}
+	}
+	p.mu.Unlock()
+
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+	select {
+	case <-done:
+	case <-time.After(timeout):
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	for _, r := range replicas {
+		r.setWaitNotify(nil)
+	}
+
+	// Make certain we don't leak the wg.Wait goroutine in the failure case.
+	// At this point, none of the callbacks will ever be called again and
+	// ch.setWaitNotify grabs a lock and so establishes the happens before.
+	all := true
+	for _, b := range caughtup {
+		if !b {
+			wg.Done()
+			all = false
+		}
+	}
+	<-done
```

Comment:
Seems odd to do this here, is the concern there's a channel leak otherwise? This will always either instantly succeed or block forever